### PR TITLE
Books login

### PR DESCRIPTION
**Pull request by eledwinn**
## Summary
1. Login page added
2. Security added, if try go to dashboard without session started, this redirect to login

## Screenshot
login

loginrefresh

## Trello Card
https://trello.com/c/svTydbMb/26-login

### CODE REVIEW

File: src/app.js [hunk 1]
```diff
@@ -7,13 +7,40 @@ angular.module('wBooksApp', [
...
   'bookDetail',
   'wInput',
-  'signUp'
+  'signUp',
+  'login'
 ]).config(['$stateProvider', '$urlRouterProvider', function($stateProvider, $urlRouterProvider) {
+  function _skipIfAuthenticated($timeout, $state, Auth) {
+    return new Promise(async (resolve, reject) => {
+      if (await Auth.isAuthenticated()) {
+        $timeout(function () {
```
Comment (FrankIglesias) [inline src/app.js]: use arrow

File: src/app.js [hunk 2]
```diff
@@ -7,13 +7,40 @@ angular.module('wBooksApp', [
...
+          $state.go('booksList');
+        });
+        reject();
+      } else {
+        resolve();
+      }
+    });
+  }
+  function _redirectIfNotAuthenticated($timeout, $state, Auth) {
```
Comment (FrankIglesias) [inline src/app.js]: use arrow

File: src/app.js [hunk 3]
```diff
@@ -7,13 +7,40 @@ angular.module('wBooksApp', [
...
+    return new Promise(async (resolve, reject) => {
+      if (await Auth.isAuthenticated()) {
+        resolve();
+      } else {
+        $timeout(function () {
```
Comment (FrankIglesias) [inline src/app.js]: use arrow

File: src/app.js [hunk 4]
```diff
@@ -7,13 +7,40 @@ angular.module('wBooksApp', [
   'booksList',
   'bookDetail',
   'wInput',
-  'signUp'
+  'signUp',
+  'login'
 ]).config(['$stateProvider', '$urlRouterProvider', function($stateProvider, $urlRouterProvider) {
+  function _skipIfAuthenticated($timeout, $state, Auth) {
```
Comment (FrankIglesias) [inline src/app.js]: arrow

File: src/core/auth/auth.service.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
+'use strict';
+
+angular.module('core.auth').factory('Auth', ['$http', 'Environment', 'Session', 'Users',
+  function ($http, Environment, Session, Users) {
```
Comment (FrankIglesias) [inline src/core/auth/auth.service.js]: arrow

File: src/core/auth/auth.service.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    const baseUrlApi = Environment.wBooksApiUrl + 'users/sessions';
+    return {
+      authenticate: (session) => {
+        return $http.post(baseUrlApi, session).then(async resp => {
+          if (resp.data.access_token) {
+            Session.create(resp.data);
+            $http.defaults.headers.common.Authorization = resp.data.access_token;
+            let user = await Users.information();
+            if (!user) {
+              user = {
```
Comment (FrankIglesias) [inline src/core/auth/auth.service.js]: why this mocked values? move to a constant aoutside

File: src/core/auth/auth.service.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+              firstName: user.first_name,
+              lastName: user.last_name,
+              email: user.email,
+              unreadNotifications: user.unread_notifications,
+              rentsCounter: user.rents_counter,
+              commentsCounter: user.comments_counter,
+              imageUrl: user.image_url
+            };
+            Session.setUser(userEntity);
+            return true;
```
Comment (FrankIglesias) [inline src/core/auth/auth.service.js]: why this true?

File: src/core/auth/auth.service.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+          } else {
+            throw new Error(resp.data.error || 'General error ocurred');
+          }
+        });
+      },
+      isAuthenticated: () => {
+        const promise = new Promise((resolve) => {
```
Comment (FrankIglesias) [inline src/core/auth/auth.service.js]: return the new promise instead

Comment (FrankIglesias) [reply src/core/auth/auth.service.js]: delete parenthesis

File: src/core/auth/auth.service.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+              email: user.email,
+              unreadNotifications: user.unread_notifications,
+              rentsCounter: user.rents_counter,
+              commentsCounter: user.comments_counter,
+              imageUrl: user.image_url
+            };
+            Session.setUser(userEntity);
+            return true;
+          } else {
+            throw new Error(resp.data.error || 'General error ocurred');
```
Comment (FrankIglesias) [inline src/core/auth/auth.service.js:25]: add a catch somewhere

File: src/login/login.component.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+          if (autenticated) {
+            $state.go('booksList');
+          }
+          this.loading = false;
+        }, error => {
+          this.loading = false;
+          alert(error);
+        });
+
+        // if (await Auth.authenticate(data)) {
```
Comment (FrankIglesias) [inline src/login/login.component.js]: remove this comments

File: src/login/login.component.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+        this.loading = true;
+
+        Auth.authenticate(data).then(autenticated => {
+          if (autenticated) {
+            $state.go('booksList');
+          }
+          this.loading = false;
+        }, error => {
+          this.loading = false;
+          alert(error);
```
Comment (FrankIglesias) [inline src/login/login.component.js]: maybe this should trigger a label

File: src/login/login.component.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        });
+
+        // if (await Auth.authenticate(data)) {
+        //   $state.go('booksList');
+        // }
+      } catch (error) {
+        // alert(error);
```
Comment (FrankIglesias) [inline src/login/login.component.js]: remove this comments

File: src/login/login.component.js [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+      }
+      // this.loading = false;
```
Comment (FrankIglesias) [inline src/login/login.component.js]: same

File: src/login/login.template.pug [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+          | Debe ingresar un email válido
+      .form-group
+        label.form-label(for="password")
+          | Password
+        input#password.form-control(type="password" name="password" ng-model="$ctrl.user.password" ng-pattern="$ctrl.passwordValidatorPattern" required)
+        p.form-validator-msg(ng-if="signUpForm.password.$error.required && signUpForm.password.$touched")
+          | Debe ingresar una contraseña
+        p.form-validator-msg(ng-if="signUpForm.password.$error.pattern")
+          | Su contraseña que tenga entre 8 y 52 caracteres, y al menos una letra y un número
+      .form-buttons
```
Comment (FrankIglesias) [inline src/login/login.template.pug:24]: check to add an error label

**Review by FrankIglesias — CHANGES_REQUESTED** [review]

File: src/core/auth/auth.service.js (hunk 6 above)
Comment (eledwinn) [reply src/core/auth/auth.service.js]: They have already been removed.

File: src/core/auth/auth.service.js (hunk 7 above)
Comment (eledwinn) [reply src/core/auth/auth.service.js]: I don't need it. Removed.

File: src/core/auth/auth.service.js (hunk 9 above)
Comment (eledwinn) [reply src/core/auth/auth.service.js:25]: Catch exists in controller.

File: src/core/auth/session.service.js [hunk 15]
```diff
@@ -0,0 +1,28 @@
+'use strict';
+
+angular.module('core.auth').factory('Session', function () {
+  const localTokenInfo = localStorage.getItem('tokenInfo');
```
Comment (FrankIglesias) [inline src/core/auth/session.service.js]: This functions can be extracted to a localStorageService

File: src/login/login.component.js [hunk 16]
```diff
@@ -0,0 +1,25 @@
+'use strict';
+
+angular.module('login').component('login', {
+  templateUrl: 'login/login.template.html',
+  bindings: { books: '<' },
+  controller: ['Auth', '$state', function BooksListController(Auth, $state) {
+    this.passwordValidatorPattern = new RegExp('^(?=.*[A-z])(?=.*[0-9]).{8,52}$');
```
Comment (FrankIglesias) [inline src/login/login.component.js]: move this regex to a constant file

**Review by FrankIglesias — CHANGES_REQUESTED** [review]

**Review by FrankIglesias — APPROVED** [review]

**Review by FrankIglesias — APPROVED** [review]

**Pull request merged**